### PR TITLE
Mednext

### PR DESCRIPTION
**Pull request by rcremese**
Fixes # .

### Description

This pull request is intended to solve some of the issues pointed by @jzielke in the MONAI pull request [8004](https://github.com/Project-MONAI/MONAI/pull/8004). A lot of the updates come from the MONAI dev branch I'm using which is ahead of your mednext branch, but let me summarize the changes I made : 
- `nets/mednext.py` 
-- renamed MedNeXt arguments to follow @jzielke [recomendations](https://github.com/Project-|MONAI/MONAI/pull/8004#discussion_r1713951167)
-- linked residual units of the encoder and decoder with one `do_res` flag
-- add model variants (S, B, M, L) as described in the original paper
-- solved code formating issues and corrected the module imports as stated in https://github.com/Project-MONAI/MONAI/blob/dev/CONTRIBUTING.md#checking-the-coding-style
- `blocks/mednext_block.py`
-- removed the author defined LayerNorm and replaced it with torch.nn.LayerNorm as only the "channel_first" case is used in block and network definition . 
-- renamed `OutBlock` to `MedNeXtOutBlock` to avoid naming confusion as the block is accessible in the API
-- fixing module class export
- `test/test_mednext.py`
-- removed all the `do_res_up_down` test cases as `do_res_up_down` as been fused with `do_res` arguments
- `__init__.py`
-- made available the different versions of MedNext to the API as done by [DenseNet](https://github.com/Project-MONAI/MONAI/blob/cea80a686b907b22d90c8024cea5a17ef9d9f58a/monai/networks/nets/densenet.py)

I should add some tests for the `mednext_block.py` script and evaluate the performance of the model on a benchmark used by the author (BTCV or AMOS22). I also tested all the basic unit test and they all pass so you don't have to do it `./runtests.sh --quick --unittests  --disttests`.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by surajpaib** [comment]
Hi @rcremese,

I think you'll need to rebase your work off of my branch so I can see the diff. changes. 

You'll need to do something like this,
```
git checkout -b surajpaib-mednext mednext
git pull git@github.com:surajpaib/MONAI.git mednext
git checkout mednext
git rebase surajpaib-mednext
```

This will generate some merge conflicts that you'll need to fix and this will give a diff. from my branch so we can compare changes between my code and yours. 

I would do it myself but I don't have push access to your branch. 

Let me know if you have any troubles with it

**Comment by surajpaib** [comment]
Alternatively, if you add me as a collaborator to your fork, I could do this myself.

**Comment by rcremese** [comment]
Oups, I guest I did something wrong. I sent you an invitation to be a contributor in my fork. I will also try to do some of the git trickery for my own and send you a new pull request.

**Comment by rcremese** [comment]
Ok ! I did what you said and hope it  will resolve the previous conflicts. Tell me if it doesn't because my git history seems grounded with commits that aren't yours.

**Comment by surajpaib** [comment]
Hi @rcremese, sorry for force-pushing on your branch but I've now updated the PR so it reflects only your changes to my code. Will do a review shortly!

File: monai/networks/nets/__init__.py [hunk 1]
```diff
@@ -53,7 +53,25 @@
 from .generator import Generator
 from .highresnet import HighResBlock, HighResNet
 from .hovernet import Hovernet, HoVernet, HoVerNet, HoverNet
-from .mednext import MedNeXt
+from .mednext import (
```
Comment (surajpaib) [inline monai/networks/nets/__init__.py:56]: @rcremese I would prefer to do a factory method pattern similar to how it exists for resnets: https://github.com/Project-MONAI/MONAI/blob/59a7211070538586369afd4a01eca0a7fe2e742e/monai/networks/nets/resnet.py#L36

for consistency

File: monai/networks/nets/mednext.py [hunk 2]
```diff
@@ -242,3 +263,131 @@ def forward(self, x: torch.Tensor) -> torch.Tensor | list[torch.Tensor]:
             return (x, *ds_outputs[::-1])
         else:
             return x
+
+
+# Define the MedNeXt variants as reported in 10.48550/arXiv.2303.09975
+class MedNeXtSmall(MedNeXt):
```
Comment (surajpaib) [inline monai/networks/nets/mednext.py:269]: Using factory methods should greatly simplify these code blocks as well!

File: monai/networks/nets/__init__.py (hunk 1 above)
Comment (rcremese) [reply monai/networks/nets/__init__.py:56]: @surajpaib, for sure I can make it look more like resnets factory method.

Comment (rcremese) [reply monai/networks/nets/__init__.py:56]: Well, I took exemple on [densnet](https://github.com/Project-MONAI/MONAI/blob/19cc6f01766120132f964beecb06d1d561f83801/monai/networks/nets/densenet.py#L152) and [SEnet](https://github.com/Project-MONAI/MONAI/blob/19cc6f01766120132f964beecb06d1d561f83801/monai/networks/nets/senet.py#L50) which are both integrated into the core library. 

Moreover, there are only 4 architecture variants in the original paper so it shouldn't polute the script that much.

Comment (surajpaib) [reply monai/networks/nets/__init__.py:56]: I see the inconsistency now. 

Factory methods are better in this case I would say - it keeps them simpler as we are not using any functionality provided by subclassing the MedNext parent. Moreover, if we want to add custom logic like loading some pre-trained weights, these lend better too.

**Comment by surajpaib** [comment]
@rcremese I'll go ahead and merge this PR - we can discuss more on my upstream

**Pull request merged**